### ISSUE TITLE
LoadClass keeps a reference to ScanResult

### ISSUE THREAD

**Issue by CDPrete** [classgraph/classgraph#399]
Loading a class with the `loadClass` method causes the loaded class to have a ClassGraph specific class loader which has still a reference to the `ScanResult` object returned by the `scan()` method.

Trying to access an inner, **non Java default**, property of the loaded classes **outside** the `use{...}` block causes **in Kotlin** an error like "The scan result cannot be access after it has been closed".

 Example:
```
// given 
enum class MyEnum { SOMETHING }

interface MyInterface {
    val myEnum: MyEnum
}

class MyClass : MyInterface {
    override val myEnum: MyEnum
        get() = SOMETHING
}

val myList = mutableListOf<MyInterface>()
ClassGraph()...config...scan().use {
    it.getClassesImplementing(MyInterface::class.java.canonicalName).forEach {
        myList.add(it.loadClass(MyInterface::class.java))
    }
}

// then, outside the use block (given that there is an element)
print(myList[0].myEnum)
```

I would expect that once loaded the class is completely unbound from ClassGraph and the ScanResult

**Comment by lukehutch** [comment]
Hmm.. I'm not sure if this can be fixed. Classloading is lazy for efficiency, so `MyEnum` isn't loaded until the field of that type is accessed. But the ClassGraph classloader is needed because the classpath or classloader list can be overridden, and accessing classpath elements can be inefficient (for example, ClassGraph may have had to extract or at least memory map and index jars within jars -- closing the `ScanResult` releases these resources).

However loading classes directly from the classpath is a last-ditch effort, if all the context classloaders failed:

https://github.com/classgraph/classgraph/blob/master/src/main/java/io/github/classgraph/ClassGraphClassLoader.java#L118

So I can probably improve the situation by making sure the `ScanResult` is not accessed unless that line is hit. That way if the context classloader can load the class, you'll never get the exception you ran into.

**Comment by lukehutch** [comment]
@CDPrete Please try the git master version, it will hopefully fix the issue in this case (although it cannot be solved in the general case, for custom classpath overrides etc., for the reasons I gave).

**Comment by CDPrete** [comment]
> @CDPrete Please try the git master version, it will hopefully fix the issue in this case (although it cannot be solved in the general case, for custom classpath overrides etc., for the reasons I gave).

Then it will fail for me since I've to override the classpath in order to indicate to the library where to find the compiled classes since I'm running it from Maven/Gradle.

What's weird is that I don't have such issue if I'm loading Java classes...

**Comment by lukehutch** [comment]
> Then it will fail for me since I've to override the classpath in order to indicate to the library where to find the compiled classes since I'm running it from Maven/Gradle.

If that's the case, you'll have to just move all code that uses classes derived from `ScanResult` into `use { ... }` (that has always been recommended anyway).

Another alternative is to "warm up" the classes inside `use { ... }` by referencing the needed field(s) in a way that won't be optimized away as a no-op. That will cause referenced classes to load.

I guess a long-term solution might be for ClassGraph to release all resources when the `ScanResult` is closed, but re-open them if the `ScanResult` is used again after it is closed. But that would break the `Autocloseable` contract.

> What's weird is that I don't have such issue if I'm loading Java classes...

My guess is that Maven/Gradle is somehow adding the classes to the classpath automatically for Java, but not for Kotlin.

Try the git master version anyway, it's possible it will improve the situation for you. I'll probably push out a new version with this change anyway, because it makes sense to make the change.

**Comment by CDPrete** [comment]
> If that's the case, you'll have to just move all code that uses classes derived from `ScanResult` into `use { ... }` (that has always been recommended anyway).
It's what I'm doing at the moment to make it work

> Another alternative is to "warm up" the classes inside `use { ... }` by referencing the needed field(s) in a way that won't be optimized away as a no-op. That will cause referenced classes to load.

So, if I understood correctly, what you're suggesting is:
1. call the `loadClass()`
2. access all the fields I need **inside** the `use{}` so that they're loaded
3. access the fields again outside where I need them for real

right? That seems really ugly and inconvenient... 

> My guess is that Maven/Gradle is somehow adding the classes to the classpath automatically for Java, but not for Kotlin.

The [kotlin-gradle-plugin](https://kotlinlang.org/docs/reference/using-gradle.html#targeting-the-jvm) should automatically attach the sources.<br/>
Moreover, in both the cases I'm overriding the classpath making the library point to where the compiled classes are written (aka target|build/classes or so).

Maven:
```
<plugin>
    <groupId>org.codehaus.mojo</groupId>
    <artifactId>exec-maven-plugin</artifactId>
    <version>1.6.0</version>
    <dependencies>
        <dependency>
            <groupId>foo.bar</groupId>
            <artifactId>foo-bar-runner</artifactId>
            <version>${foo-bar.version}</version>
        </dependency>
    </dependencies>
    <executions>
        <execution>
            <phase>process-classes</phase> 
            <goals>
                <goal>java</goal>
            </goals>
            <configuration>
                <mainClass>foo.bar.MyRunner</mainClass> 
                <cleanupDaemonThreads>false</cleanupDaemonThreads>
                <arguments>
                    <argument>${project.build.outputDirectory}</argument>  
                    <argument>${project.build.directory}</argument>    
                </arguments>
            </configuration>
        </execution>
    </executions>
</plugin>
```

Gradle:
```
configurations {
        runner
}
dependencies {
        asciidocModuleCodes "foo.bar:foo-bar-runner:${fooBarVersion}"
}
task generateAsciidocFromModuleCodes(type: JavaExec) {
        classpath = configurations.runner 
        main = 'foo.bar.MyRunner'  
        args sourceSets.main.java.outputDir, project.buildDir  
        dependsOn compileKotlin
    }
```

Is there something wrong?

**Comment by lukehutch** [comment]
> right? That seems really ugly and inconvenient...

Yes, it is... but once temporary resources have been cleaned up and files unmapped, all that would have to be undone again to get access to jars every time a class was loaded, which would be inefficient. Leaving the resources open causes a resource leak, and there have been numerous bugs about resource leaks in the past (when ClassGraph used finalizers and/or shutdown hooks to clean up resources), which is why ClassGraph now requires a `use { ... }` or try-with-resources to close the ScanResult. (ClassGraph uses its own sophisticated zipfile central directory parser to allow for highly parallelized access to jars and jars-within-jars, and this uses memory mapping, so these resources have to be properly closed when no longer needed.)

However all is not lost. I just checked in a change that should probably solve the problem for your usecase. When `ClassGraphClassLoader` is instantiated, it gets the classpath as a list of URLs, and sets up a `URLClassLoader` based on these classpath URLs. If environment classloaders can't load the class, and/or the classpath has been overridden, then classloading will be delegated to this `URLClassLoader`. Java's own internal caching of open jarfiles can be relied on for resource management. This is less efficient than using ClassGraph's zipfile management, and is not parallelizable (there's a single-threaded lock around Java's zipfile API), but should work even after the `ScanResult` has been closed. Can you please test the git master version?

> Is there something wrong?

Honestly I know very little about how Maven works under the hood, and even less about Gradle, but I suspect if you printed out the classpath in Maven and Gradle for both the Java and Kotlin usecases, you would see a difference.

**Comment by lukehutch** [comment]
I'm pretty sure this should be fixed for you in 4.8.62, which I just released. Please test and report back, and I'll re-open the bug if you still experience the problem. Thanks!

**Comment by lukehutch** [comment]
Release notes here:

https://github.com/classgraph/classgraph/releases/tag/classgraph-4.8.62
